composer/loggers/neptune_logger.py

```diff
@@ -0,0 +1,257 @@
+        upload_artifacts (bool, optional): Whether the logger should upload artifacts to Neptune. Default: ``False``.
+        base_namespace (str, optional): The name of the base namespace to log the metadata to. Default: "training".
+        neptune_kwargs (Dict[str, Any], optional): Any additional keyword arguments to the neptune.init_run()
+            function. For options, see the [Run API reference](https://docs.neptune.ai/api/neptune/#init_run) in the
+            Neptune docs.
+
+    Example:
+        Default Neptune logger:
+        >>> from composer.loggers import NeptuneLogger
+        >>> neptune_logger = NeptuneLogger()
+        >>> trainer = Trainer(loggers=neptune_logger, ...)
+        >>> trainer.fit()
+
+        More options:
+        >>> neptune_logger = NeptuneLogger(
+        ...     project="ml-team/classification",
+        ...     upload_artifacts=True,
+        ...     dependencies="infer",  # Neptune Run kwarg
+        )
+        >>> trainer = Trainer(loggers=neptune_logger, ...)
+        >>> trainer.fit()
+
+    For more, see the [Neptune-Composer integration guide](https://docs.neptune.ai/integrations/composer/).
+    """
+    METRIC_NAMESPACE = 'metrics'
+    HYPERPARAM_NAMESPACE = 'hyperparameters'
+    TRACE_NAMESPACE = 'traces'
+    INTEGRATION_VERSION_KEY = 'source_code/integrations/neptune-MosaicML'
+
+    def __init__(
+        self,
+        *,
+        project: Optional[str] = None,
+        api_token: Optional[str] = None,
+        rank_zero_only: bool = True,
+        upload_artifacts: bool = False,
+        base_namespace: str = 'training',
+        **neptune_kwargs,
+    ) -> None:
+        try:
+            from neptune import Run
+            from neptune.internal.utils import verify_type
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='neptune',
+                                                conda_package='neptune',
+                                                conda_channel='conda-forge') from e
+
+        verify_type('project', project, (str, type(None)))
+        verify_type('api_token', api_token, (str, type(None)))
+        verify_type('rank_zero_only', rank_zero_only, bool)
+        verify_type('upload_artifacts', upload_artifacts, bool)
+        verify_type('base_namespace', base_namespace, str)
+
+        if not base_namespace:
+            raise ValueError("Argument 'base_namespace' cannot be an empty string.")
+
+        self._project = project
+        self._api_token = api_token
+        self._rank_zero_only = rank_zero_only
+        self._upload_artifacts = upload_artifacts
+        self._base_namespace = base_namespace
+        self._neptune_kwargs = neptune_kwargs
+
+        self._enabled = (not rank_zero_only) or dist.get_global_rank() == 0
+        self._neptune_run = None
+        self._base_handler = None
+
+        if self._enabled:
+            self._neptune_run = Run(
```

Comment:
A few other integrations showed us that it's quite unsafe to initialize a Run in class constructor. Let's made it lazy evaluated in `neptune_run` property. And let's just use a Run in a `debug` mode to prevent the whole logic of `if self._neptune_run is not None:...`